src/lib.rs

```diff
@@ -146,16 +146,67 @@ fn get_num_cpus() -> usize {
     cpus as usize
 }
 
-#[cfg(
-    any(
-        target_os = "linux",
-        target_os = "nacl",
-        target_os = "macos",
-        target_os = "ios",
-        target_os = "android",
-        target_os = "solaris",
-        target_os = "fuchsia",
-    )
+#[cfg(feature = "nightly")]
+#[feature(asm, naked_functions)]
+#[naked]
+#[inline(always)]
+#[cfg(all(target_os = "linux", target_arch = "x86_64"))]
+fn popcnt(bits: u64) -> u64 {
+    let ret: u64;
+    unsafe {
+        asm!("popcntq %rdi, %rax"
+             : "={rax}"(ret)
+             : "{rdi}"(bits)
+        );
+    };
+    ret
+}
+
+#[cfg(not(feature = "nightly"))]
+#[inline(always)]
+#[cfg(all(target_os = "linux", target_arch = "x86_64"))]
+fn popcnt(n: u64) -> u64 {
+    let mut c: u64 = (n & 0x5555555555555555) + ((n >> 1) & 0x5555555555555555);
+    c = (c & 0x3333333333333333) + ((c >> 2) & 0x3333333333333333);
+    c = (c & 0x0f0f0f0f0f0f0f0f) + ((c >> 4) & 0x0f0f0f0f0f0f0f0f);
+    c = (c & 0x00ff00ff00ff00ff) + ((c >> 8) & 0x00ff00ff00ff00ff);
+    c = (c & 0x0000ffff0000ffff) + ((c >> 16) & 0x0000ffff0000ffff);
+    c = (c & 0x00000000ffffffff) + ((c >> 32) & 0x00000000ffffffff);
+    c
+}
+
+#[inline(always)]
+#[cfg(all(target_os = "linux", not(target_arch = "x86_64")))]
+fn popcnt(n: u64) -> u64 {
+    let mut c: u64 = (n & 0x5555555555555555) + ((n >> 1) & 0x5555555555555555);
+    c = (c & 0x3333333333333333) + ((c >> 2) & 0x3333333333333333);
+    c = (c & 0x0f0f0f0f0f0f0f0f) + ((c >> 4) & 0x0f0f0f0f0f0f0f0f);
+    c = (c & 0x00ff00ff00ff00ff) + ((c >> 8) & 0x00ff00ff00ff00ff);
+    c = (c & 0x0000ffff0000ffff) + ((c >> 16) & 0x0000ffff0000ffff);
+    c = (c & 0x00000000ffffffff) + ((c >> 32) & 0x00000000ffffffff);
+    c
+}
+
+#[cfg(target_os = "linux")]
+fn get_num_cpus() -> usize {
```

Comment:
I wonder whether using affinity by default is better, or if it'd be a sort of breaking change. Would one always always want the affinity, or could one conceivably want to just know the number or CPUs on the machine anyways?

Put another way, should this be a separate function, or just be the default and only implementation for linux? I'm not saying one or the other is correct, I'd be interested in rationale for either case.